File: LEETCODE/EASY/DAY 18/PLUS ONE.py
from typing import List

class Solution:
    def plusOne(self, digits: List[int]) -> List[int]:
        for i in range(len(digits) - 1, -1, -1):
            if digits[i] < 9:
                digits[i] += 1
                return digits
            digits[i] = 0

        # If all were 9s
        return [1] + digits


# 🚀 Example Run
sol = Solution()
print(sol.plusOne([1,2,3]))   # [1,2,4]
print(sol.plusOne([9,9,9]))   # [1,0,0,0]

File: LEETCODE/EASY/DAY 18/PLUS ONE.java
import java.util.*;

public class PlusOne {
    public List<Integer> plusOne(List<Integer> digits) {
        for (int i = digits.size() - 1; i >= 0; i--) {
            int digit = digits.get(i);

            if (digit < 9) {
                digits.set(i, digit + 1);
                return digits;
            } else {
                digits.set(i, 0);
            }
        }

        // If all digits were 9
        digits.add(0);  
        digits.set(0, 1);
        return digits;
    }

    public static void main(String[] args) {
        PlusOne sol = new PlusOne();
        System.out.println(sol.plusOne(new ArrayList<>(Arrays.asList(1,2,3)))); // [1,2,4]
        System.out.println(sol.plusOne(new ArrayList<>(Arrays.asList(9,9,9)))); // [1,0,0,0]
    }
}
